[Infrastructure] Adds support for capturing process dumps for hanging builds on Windows

javiercn opened the pull request:
* Downloads and installs ProcDump as part of the build.
* Registers a scheduled job that wakes up in 160 minutes.
* Upon waking up, the background job lists all the processes it
  received as candidates.
* For each process it captures a full memory dump.
* At the end of the build, a separate step checks on whether the job
  ran (the build hanged) or everything is correct and displays statistics.
* If dumps are found, they are collected and made available as artifacts
  under Windows(_Templates)_Tests_Dumps.

A future improvement is to use `dotnet-dump` instead of ProcDump to cover Linux too, but starting with ProcDump as it also allows capturing dumps on First Chance Exceptions, which is something I'm also looking to enable.

javiercn committed:
[Infrastructure] Adds support for capturing process dumps for hanging builds on Windows.
* Downloads and installs ProcDump as part of the build.
* Registers a scheduled job that wakes up in 160 minutes.
* Upon waking up, the background job lists all the processes it
  received as candidates.
* For each process it captures a full memory dump.
* At the end of the build, a separate step checks on whether the job
  ran (the build hanged) or everything is correct and displays statistics.
* If dumps are found, they are collected and made available as artifacts
  under Windows(_Templates)_Tests_Dumps.

javiercn commented:
/azp run

dougbu commented on eng/scripts/FinishDumpCollectionForHangingBuilds.ps1:
```diff
@@ -0,0 +1,74 @@
+  [Parameter(Mandatory = $true)]
+  [ValidateNotNullOrEmpty()]
+  [string]
+  $ProcDumpDumpFolderPath
```
Suggest `$ProcDumpOutputPath`

dougbu commented on eng/scripts/FinishDumpCollectionForHangingBuilds.ps1:
```diff
@@ -0,0 +1,74 @@
+
+$sentinelFile = Join-Path $ProcDumpDumpFolderPath "dump-sentinel.txt";
+if ((-not (Test-Path $sentinelFile))) {
+  Write-Output "No sentinel file available in '$sentinelFile'.";
```
Suggest something like "StartDumpCollectionForHangingBuilds.ps1 has not been executed".

dougbu commented on eng/scripts/StartDumpCollectionForHangingBuilds.ps1:
```diff
@@ -0,0 +1,112 @@
+  New-Item -ItemType Directory $ProcDumpDumpFolderPath;
+}
+
+# We write a sentinel file that we use at the end of the build to determine if there were build errors or the build hanged.
```
Since the file is never deleted, it can't be used to "determine if there were build errors or the build hanged".

dougbu commented on eng/scripts/FinishDumpCollectionForHangingBuilds.ps1:
```diff
@@ -0,0 +1,74 @@
+Write-Host "Listing existing jobs";
+Get-Job
+Write-Host "Listing existing scheduled jobs";
+Get-ScheduledJob
```
Why is this output useful?

javiercn replied:
Debug purposes. Registering the Job is trickier than it seems.

dougbu replied:
Now you've figured it out, listing all jobs on the agent will just be confusing. Please remove or at least limit the output to the `CaptureDumps*` jobs.

javiercn replied:
Same reason as for the other comment. This information was useful in troubleshooting issues, its useful to keep it here for troubleshooting future issues.

I don't see how this is confusing, the output clearly states what the information being displayed is and if you don't think you need it, you can simply skip over it.

dougbu reviewed (changes requested):
Main issue is direct download of the Zip file

javiercn commented on eng/scripts/InstallProcDump.ps1:
```diff
@@ -0,0 +1,46 @@
+Write-Host "Done downloading ProcDump"
+Expand-Archive "$tempDir/ProcDump.zip" -d "$tempDir/ProcDump/"
+Write-Host "Expanded ProcDump to $tempDir"
+Write-Host "Installing ProcDump to $installDir"
```
Follows the same pattern InstallJavaSDK does. Don't want to do something different without a good reason.

javiercn commented on eng/scripts/StartDumpCollectionForHangingBuilds.ps1:
```diff
@@ -0,0 +1,112 @@
+# The dumps are placed in $ProcDumpDumpFolderPath
+# If the build completes sucessfully in less than two hours, a final step unregisters the job.
+
+# Create a unique identifier for the job name
```
I've seen conflicts happen if the job doesn't get cleaned up properly.
Using a unique identifier avoids the issue.
We don't want the build to fail or skip capturing dumps for potential hanging builds just because a previous job didn't got cleaned up properly.
Id rather be err on the side of reliability here.

javiercn committed:
Addressed feedback

javiercn commented on eng/scripts/InstallProcDump.ps1:
```diff
@@ -0,0 +1,46 @@
+mkdir $tempDir -ea Ignore | out-null
+mkdir $installDir -ea Ignore | out-null
+Write-Host "Starting ProcDump download"
+Invoke-WebRequest -UseBasicParsing -Uri "https://download.sysinternals.com/files/Procdump.zip" -Out "$tempDir/ProcDump.zip"
```
We do this in several places in the pipeline so I don't want to tackle that as part of this PR as it will unnecessarily slow it down and it can be done in parallel. Instead, I filed an issue to fix this for all the places where we download an external asset. See https://github.com/aspnet/AspNetCore-Internal/issues/3125 for details

dougbu replied:
Glad you filed the issue. Have you started the conversation w/ the Arcade folks about making procDump available?

javiercn replied:
I expect the issue to follow the normal triage process and that action is taken when it gets scheduled based on priority.

dougbu commented on eng/scripts/StartDumpCollectionForHangingBuilds.ps1:
```diff
@@ -0,0 +1,123 @@
+
+# Ensure that the dumps output path exists.
+if ((-not (Test-Path $ProcDumpOutputPath))) {
+  New-Item -ItemType Directory $ProcDumpOutputPath;
```
`| out-null` to remove useless logging

dougbu commented on eng/scripts/StartDumpCollectionForHangingBuilds.ps1:
```diff
@@ -0,0 +1,123 @@
+$ScriptTrigger = New-JobTrigger -Once -At $WakeTime;
+
+try {
+  Register-ScheduledJob -Name $JobName -ScriptBlock $ScriptCode -Trigger $ScriptTrigger -ArgumentList $ProcDumpPath, $ProcDumpOutputPath, $CandidateProcessNames;
```
This writes a lot of information to the CI console, very little of which matters even when debugging. Suggest `| out-null` or perhaps `Format-list {property that matters}, ...` (where the number of properties is low and definitely does not include the script)

javiercn replied:
What's the problem with that? Id rather see more information than less when we need to debug scripts. This doesn't cause any harm and it has obvious advantages in the case that in the future something goes wrong. (We captured all the information at the time, and we don't have to go and modify things to capture information in the future). Based on that, I'm leaving the script as is.

dougbu replied:
Well, something changed here 😃 and the output is much more compact. Thanks.

dougbu commented on .azure/pipelines/ci.yml:
```diff
@@ -431,6 +431,9 @@ stages:
         publishOnError: true
+      - name: Windows_Test_Dumps
+        path: artifacts/dumps/
+        publishOnError: true
```
Suggest making the dump output location artifacts/log and reverting changes to this file. That's where dump_process.ps1 output already goes.

javiercn replied:
Dumps are not logs, id rather keep them separate.

dougbu reviewed (commented):
One

javiercn committed:
Address additional feedback

dougbu commented:
@javiercn the general point is you seem to be writing this for yourself. Even then, displaying information about unrelated jobs and showing the script that's right in the .ps1 file don't help.

My point about artifact/log is we already have too many artifacts, making it hard to find anything and a non-required artifact containing only one or two files isn't enough to justify its existence.

javiercn committed:
Address more feedback

dougbu commented on .azure/pipelines/jobs/default-build.yml:
```diff
@@ -126,6 +126,11 @@ jobs:
+    - ${{ if and(eq(parameters.agentOs, 'Windows'), eq(parameters.isTestingJob, true)) }}:
+      - powershell: ./eng/scripts/InstallProcDump.ps1
+        displayName: Install ProcDump
+      - powershell: ./eng/scripts/StartDumpCollectionForHangingBuilds.ps1 $(ProcDumpPath)procdump.exe artifacts/log/ (Get-Date).AddMinutes(160) dotnet
```
Thanks

dougbu commented on eng/scripts/FinishDumpCollectionForHangingBuilds.ps1:
```diff
@@ -0,0 +1,76 @@
+}
+
+Write-Output "Listing existing jobs";
+Get-Job
```
` -Name CaptureDumps*` here too?

javiercn commented:
> the general point is you seem to be writing this for yourself

I'm not even sure how to answer this. I really don't know what you mean.

> @javiercn the general point is you seem to be writing this for yourself. Even then, displaying information about unrelated jobs and showing the script that's right in the .ps1 file don't help.

I've narrowed it down to show just jobs that match the pattern CaptureDump*.

> My point about artifact/log is we already have too many artifacts, making it hard to find anything and a non-required artifact containing only one or two files isn't enough to justify its existence.

I don't agree with this. Putting the artifact in Logs makes it harder to find, simply because it makes the artifact visible when you click in the artifacts tab. That said, I changed it because I care way more about us capturing this information than about us discussing where it goes.

I think I've addressed all your feedback at this point.

javiercn committed:
Final feedback
